### PR TITLE
[#261] Table MapsCodec

### LINKED ISSUE #261: Table Maps Codec
This is a follow up to #254. #255 Solved the problem of dynamic keys if the values are simple values. But I need the values to be more complex than that. I need a way of specifying a codec for the values.

### PR DESCRIPTION
**Pull request by chshersh**
Resolves #261

@vrom911 and I brainstormed the solution for this problem and came up with a nice implementation. I'm quite happy that we could avoid full-batteries TOML codec running. @everythingfunctional you can check it out, it should address your use case that you've attempted to solve in #262.

### CODE REVIEW

File: src/Toml/Codec/Combinators.hs [hunk 1]
```diff
@@ -516,34 +520,112 @@ map keyCodec valCodec key = Codec input output
...
-                case backward valBiMap anyVal of
-                    Right v  -> case forward keyBiMap tomlKey of
-                        Right k  -> pure (k, v)
-                        Left err -> throwError $ BiMapError err
+    input :: TomlEnv (Map k v)
+    input = asks (Prefix.lookup tableName . tomlTables) >>= \case
+        Nothing -> pure Map.empty
+        Just toml -> local (const toml) $ do
+            valKeys <- asks (HashMap.keys . tomlPairs)
+            tableKeys <- asks (fmap (:|| []) . HashMap.keys . tomlTables)
```
Comment (everythingfunctional) [inline src/Toml/Codec/Combinators.hs:615]: Do you also need to add this?
```haskell
tableArrayKeys <- asks (HashMap.keys . tomlTableArrays)
```

File: src/Toml/Codec/Combinators.hs [hunk 2]
```diff
@@ -516,34 +520,112 @@ map keyCodec valCodec key = Codec input output
...
+    { configColours :: 'Map' 'Text' Colour
+    }
+@
+
+And you want in the result to have a 'Map' like this:
+
+@
+'Map.fromList'
+    [ "yellow" -> Hex "#FFFF00"
+    , "ping"   -> Hex "#FFC0CB"
```
Comment (everythingfunctional) [inline src/Toml/Codec/Combinators.hs:565]: typo here. "ping" should be "pink"

**Comment by everythingfunctional** [comment]
This does look much better. I clearly still have a lot to learn about Haskell programming. And I'm pretty sure it does solve my problem. Thanks.

File: src/Toml/Codec/Combinators.hs (hunk 1 above)
Comment (vrom911) [reply src/Toml/Codec/Combinators.hs:615]: Don't think that we should support such structures.

Comment (everythingfunctional) [reply src/Toml/Codec/Combinators.hs:615]: Why not? I can't think of a use case off the top of my head, but I'm just curious.

Comment (chshersh) [reply src/Toml/Codec/Combinators.hs:615]: By design, TOML the language discourages nested data structures. If there are compelling use-cases, we can consider adjusting the implementation.

Comment (everythingfunctional) [reply src/Toml/Codec/Combinators.hs:615]: Ok, fair enough. Thanks.

**Pull request merged**